hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastIntegerType.java

```diff
@@ -0,0 +1,421 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.types;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.type.BasicSqlType;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.calcite.sql.type.SqlTypeName.BIGINT;
+import static org.apache.calcite.sql.type.SqlTypeName.INTEGER;
+import static org.apache.calcite.sql.type.SqlTypeName.SMALLINT;
+import static org.apache.calcite.sql.type.SqlTypeName.TINYINT;
+
+/**
+ * Represents integer-valued types for Calcite.
+ * <p>
+ * Unlike the standard Calcite implementation for TINYINT, SMALLINT, INT and
+ * BIGINT, this implementation tracks the actual bit width required to represent
+ * integer values. That bit width information is not directly related to the
+ * underlying machine representation. The closest concept is SQL precision
+ * tracking done for DECIMAL type, but the bit width information tracks the
+ * precision in terms of binary (base-2) digits instead of decimal (base-10)
+ * digits. For instance, -1 and 1 require a single binary digit, 14 requires 4
+ * binary digits.
+ * <p>
+ * In general, for an N-bit integer type the valid range of bit widths is from 0
+ * to N: zero bit width corresponds to 0 integer value, bit widths from 1 to
+ * N - 1 correspond to regular integer values, bit width of N bits has a special
+ * meaning and indicates a possibility of an overflow.
+ * <p>
+ * For instance, for BIGINT type represented as Java {@code long} type:
+ * the valid bit width range is from 0 to 64, 0L has BIGINT(0) type, -14L has
+ * BIGINT(4) type, a BIGINT SQL column has BIGINT(63) type, BIGINT(64) indicates
+ * a potential overflow.
+ * <p>
+ * Each arithmetic operation acting on integer types infers its return type
+ * based on the bit width information: INT(4) + INT(10) -> INT(11),
+ * INT(4) + INT(31) -> BIGINT(32), INT(10) + BIGINT(63) -> BIGINT(64). In the
+ * first example the bit width was expanded; in the second example the bit width
+ * was expanded and the type was promoted to the next wider integer type; in the
+ * last example the bit width was expanded, but there was no wider integer type
+ * to promote to and avoid a possible overflow.
+ * <p>
+ * The benefits of that bit width approach are: the smallest possible type is
+ * always selected to represent a result of a certain operation and it's always
+ * possible to tell from the selected type alone whether overflow checking is
```

Comment:
Is this "no-overflow checking" integrated with expressions somehow? For example, in the PLUS function, we always check for overflow, even for BIGINT(62), where the overflow is impossible.

Reply:
explained here https://github.com/hazelcast/hazelcast/pull/17183#discussion_r452824415